### PR TITLE
[MRG] Test refactoring

### PR DESCRIPTION
**Pull request by jaeilepp**
Closes https://github.com/mne-tools/mne-python/issues/2151.
I made a function for testing different kind of readers. The filtering precision probably depends on the data a lot, so I'm not sure if it should be included here or tested separately for each reader. What do you think? For now it is implemented only for nicolet reader, but I will implement for others too once we have tuned this and if you think it's useful.

Also included a small optimization for ``test_epochs``.

### CODE REVIEW

**Comment by agramfort** [comment]
objective is to:

- reduce test time
- remove test lines of code via common tests for all Raw objects.

then whatever works !

**Comment by jaeilepp** [comment]
I made the filter test it's own private function. This way the precision can be set case by case.

**Comment by agramfort** [comment]
ok. Can you factorize the tests of other readers?

File: mne/io/nicolet/tests/test_nicolet.py [hunk 1]
```diff
@@ -18,26 +16,5 @@
...
-    data1, times1 = raw[:10:3, 10:12]
-    data2, times2 = raw2[:10:3, 10:12]
-    data3, times3 = raw2[[0, 3, 6, 9], 10:12]
-    assert_array_almost_equal(data1, full_data[:10:3, 10:12], 9)
-    assert_array_almost_equal(data1, data2, 9)
-    assert_array_almost_equal(data1, data3, 9)
-    assert_array_almost_equal(times1, times2)
-    assert_array_almost_equal(times1, times3)
+    raw = _test_raw_object(read_raw_nicolet, fname)
+    _test_raw_filter(raw, precision=5)
```
Comment (Eric89GXL) [inline mne/io/nicolet/tests/test_nicolet.py]: usually `assert_allclose` is better than using `assert_array_almost_equal` because it's better at dealing with different magnitudes of data, so having `atol` and `rtol` arguments is better here

**Comment by jaeilepp** [comment]
Ok. I've refactored the tests for all the possible cases. Are we happy with this? Did I miss anything or remove too much?

**Comment by Eric89GXL** [comment]
Looks like more red than green, which is good. Can you comment on test time on `master` compared to this branch for `mne/io/`? Looks like coverage made a tiny increase which is good.

File: mne/io/edf/tests/test_edf.py [hunk 2]
```diff
@@ -52,8 +53,8 @@ def test_concat():
 
 def test_bdf_data():
     """Test reading raw bdf files"""
-    raw_py = read_raw_edf(bdf_path, montage=montage_path, eog=eog,
-                          misc=misc, preload=True)
+    raw_py = _test_raw_object(read_raw_edf, True, input_fname=bdf_path,
```
Comment (Eric89GXL) [inline mne/io/edf/tests/test_edf.py]: I assume `True` is `preload=True` but it's probably better to enter as `kwarg`

Comment (jaeilepp) [inline mne/io/edf/tests/test_edf.py]: It's actually a boolean telling whether not preloading is implemented for the reader. So basically if it's true both cases are tested.

Comment (Eric89GXL) [inline mne/io/edf/tests/test_edf.py]: Ahh, it's for testing preloading. I wonder which readers still don't have preloading...

Comment (Eric89GXL) [inline mne/io/edf/tests/test_edf.py]: Or non-preloading (access on demand) I mean

**Comment by Eric89GXL** [comment]
Looks like the right direction. Two comments:

1. For many (all?) readers it's useful to be able to pass a set of time slices to check, usually around the boundaries of "blocks" within a given file. For an example see [here] (https://github.com/mne-tools/mne-python/pull/2628/files#diff-89350124c3b525af7b8aabdf157ae7ecR131).

2. I don't see `fiff/tests/test_raw.py` refactored here. I assume that's because you're waiting to do it. But a lot of our core `_BaseRaw` functionality testing actually occurs in `fiff/tests/test_raw.py`. In other words, `RawFIFF` is now almost as simple as our other readers (IIRC it adds split-file detection, and maxshield options, and not much else). Since we are now testing e.g. `filtering` for all our Raw types, we probably want to move almost all of the code from `fiff/tests/test_raw.py` to your new location.

File: mne/io/edf/tests/test_edf.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/io/edf/tests/test_edf.py]: Okay. Keep in mind there are actually there `preload` modes, not two, though: False (on demand), True (in memory) and str (memmapping), so you probably want to add a third. This is one of those that is probably most completely tested in `fiff/tests/test_raw.py`.

Comment (jaeilepp) [inline mne/io/edf/tests/test_edf.py]: Actually memmapping is also tested if True.

**Comment by jaeilepp** [comment]
> Looks like more red than green, which is good. Can you comment on test time on master compared to this branch for mne/io/? Looks like coverage made a tiny increase which is good.

I don't see a big difference in test times. I suspect it may even go up with this, since filtering was not tested in most readers before

**Comment by jaeilepp** [comment]
Seems that it takes 4 seconds longer than on master (64s for all the io tests vs. 60s on master).

**Comment by agramfort** [comment]
maybe we don't need to test filtering for all classes but test it at the
level of the base class ... I am just thinking out loud.

reducing testing time would be awesome...

**Comment by Eric89GXL** [comment]
Most reading classes only need to implement `__init__` and `_read_segment_file` (and that only if supporting on-demand reads), so it *should* suffice to check data extraction using the `raw[:]` interface. As long as we check that thoroughly (e.g., data values, dtypes, maybe concatenation) I think we should be able to test filtering in one place. @jaeilepp could you make a list of what all needs to be tested so we can see if it makes sense? Here's what I can think of offhand:

Reader-specific:

1. Data reading equivalence with mne-C under preload conditions
2. Verify data with `proj=True` and `proj=False` against mne-C-read versions (ideally all should support projection, not sure if all do)
3. Reading different slices of data, especially near "block" boundaries in the original format.
4. Anything format specific (e.g., additional args / modes)

Reader-generic:

1. Filtering
2. Concatenation
3. ...?

**Comment by jaeilepp** [comment]
I think if we only test concatenation and filtering for the base class, the test for readers would only contain slicing, preloading, reading and writing + the reader specific functions. Do you mean that the data should also be tested against the mne-c conversion functions (kit2fiff etc.)?

I think all the generic functions are (or should be) already tested in the test_raw module.

**Comment by jaeilepp** [comment]
I think at the moment there is a lot of redundancy with the reader tests (concatenation, filtering etc.). So by only testing the base class, the test times should go down.

**Comment by agramfort** [comment]
Same feeling here

File: mne/io/tests/test_raw.py [hunk 3]
```diff
@@ -1,14 +1,71 @@
...
+    Returns
+    -------
+    raw : Instance of Raw
+        A preloaded Raw object.
+    """
+    tempdir = _TempDir()
+    raws = list()
+    raws.append(reader(**kwargs))
+    picks = [1, 3, 5]
+    bnd = 100
```
Comment (jaeilepp) [inline mne/io/tests/test_raw.py]: Not sure what would be the best way to figure out the boundary. Is there some kind of generic tag for block boundaries in the info structure or somewhere else? Or should we just add a function argument?

Comment (agramfort) [inline mne/io/tests/test_raw.py]: no idea. AFAIK only fif and certain CTF files store data in buffers.

Comment (Eric89GXL) [inline mne/io/tests/test_raw.py]: And EDF. The reader should pass the boundary as an arg, see the link I sent
to my PR above for the type of tests you can do

Comment (Eric89GXL) [inline mne/io/tests/test_raw.py]: Actually raw.info['buffer_size_sec'] should have it

Comment (Eric89GXL) [inline mne/io/tests/test_raw.py]: So if you do `int(round(raw.info['buffer_size_sec'] * raw.info['sfreq']))` you have the block borders. And for readers that don't have blocks, you can have `test_blocks=False` by default in your reading function test. Make sense?

Comment (Eric89GXL) [inline mne/io/tests/test_raw.py]: Look here for code you can copy-paste:

https://github.com/mne-tools/mne-python/pull/2628/files#diff-89350124c3b525af7b8aabdf157ae7ecR145

The two lines that are reader-specific here are the `assert_equal` lines involving `bnd`, the others could (and probably should) be used for any reader.

Comment (agramfort) [inline mne/io/tests/test_raw.py]: Sounds good to me

**Comment by jaeilepp** [comment]
While implementing the slicing tests, I failed to make it work for brainvision reader. It fails on ``slice(bnd - 1, bnd)``. I'm starting to wonder if it's a bug in the reader.

Also noticed that for edf files the picks must be sorted.

**Comment by agramfort** [comment]
>
> Also noticed that for edf files the picks must be sorted.
>
it should not be the case...

**Comment by Eric89GXL** [comment]
It's possible the brainvision reader has a bug. If it has blocks of data,
refactoring to use _block_read_limits from my just merged PR should fix it.
I can take a look if you want. The picks sorting could also be a bug.

**Comment by Eric89GXL** [comment]
Thinking about how the EDF reader works they do need to be sorted. It can
be fixed by using fid.seek(pos, 0) appropriately like in kit. I can also
fix that if you want. It'll be a little bit trickier because each channel
can have a different number of samples, but a cumsum should give the
offsets nicely.

**Comment by Eric89GXL** [comment]
I meant "like in CTF", not "kit".

**Comment by jaeilepp** [comment]
I can also look at it if you're busy. Depends on how urgent this is. If you want it done quickly you should probably do it :)

**Comment by Eric89GXL** [comment]
If you can rebase if you haven't already and leave EDF and BV failing I can
look today or tomorrow sometime

**Comment by Eric89GXL** [comment]
PR sent

**Comment by Eric89GXL** [comment]
Hooray green!

**Comment by agramfort** [comment]
@Eric89GXL are you happy here?

**Comment by Eric89GXL** [comment]
Is it ready for review then?

**Comment by jaeilepp** [comment]
Well, there is still some redundancy in my opinion. For instance, edf and kit tests still test concatenation by calling ``_test_concat`` in ``mne.io.tests.test_raw``. Are we okay removing them and only testing this function for fiff-class?

**Comment by Eric89GXL** [comment]
Now that I think about it, concatenation tests ensure that `first_samp` and `last_samp` are set appropriately, so maybe we want all readers to be tested for concatenation?

**Comment by agramfort** [comment]
>
> Now that I think about it, concatenation tests ensure that first_samp and
> last_samp are set appropriately, so maybe we want all readers to be
> tested for concatenation?
>
fine with me

**Comment by jaeilepp** [comment]
Merged https://github.com/mne-tools/mne-python/pull/2655.
Is the concat test sufficient?

**Comment by jaeilepp** [comment]
It seems kit and edf are still using ``_test_concat`` function. Are you ok removing these and only testing concatenation thoroughly with fiff?

**Comment by agramfort** [comment]
if we can guarantee correctness otherwise yes.

**Comment by agramfort** [comment]
@jaeilepp make PR as MRG when ready.

make sure you don't remove test that are not tested otherwise.

**Comment by jaeilepp** [comment]
Added a couple of small tests just to be on the safe side. I think this should be good now.

File: mne/tests/test_epochs.py [hunk 4]
```diff
@@ -414,16 +414,13 @@ def test_read_write_epochs():
...
 
         # decim without lowpass
-        lowpass = raw.info['lowpass']
-        raw.info['lowpass'] = None
-        epochs_dec = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
-                            baseline=(None, 0), decim=4)
+        epochs_dec.info['lowpass'] = None
+        epochs_dec.decimate(2)
         assert_equal(len(w), 2)
-        raw.info['lowpass'] = lowpass
```
Comment (agramfort) [inline mne/tests/test_epochs.py]: why did you change this?

**Comment by agramfort** [comment]
do you test that highpass and lowpass and consistent between readers?

**Comment by agramfort** [comment]
did you gain any test time?

**Comment by Eric89GXL** [comment]
And please make sure no new un-caught warnings pop up. I'm pretty sure that `lowpass`-setting line that @agramfort asked about above was done to avoid an uncaught warning or induce a caught one.

File: mne/io/nicolet/tests/test_nicolet.py [hunk 5]
```diff
@@ -5,11 +5,9 @@
...
 import inspect
-from numpy.testing import assert_array_almost_equal, assert_array_equal
 
-from mne import concatenate_raws
-from mne.io import read_raw_nicolet, Raw
-from mne.utils import _TempDir
+from mne.io import read_raw_nicolet
+from mne.io.tests.test_raw import _test_raw_object
 
 FILE = inspect.getfile(inspect.currentframe())
```
Comment (Eric89GXL) [inline mne/io/nicolet/tests/test_nicolet.py]: instead of doing this we usually do `file = op.dirname(__file__)`

**Comment by Eric89GXL** [comment]
How does test time look on your system? Otherwise the changes look reasonable to me.

**Comment by jaeilepp** [comment]
Epochs test throws no warnings for me, but I can put it back. Shouldn't make any difference.

**Comment by jaeilepp** [comment]
> do you test that highpass and lowpass and consistent between readers?

Didn't we agree to test filtering only for fiff?

**Comment by Eric89GXL** [comment]
>> do you test that highpass and lowpass and consistent between readers?

> Didn't we agree to test filtering only for fiff?

IIUC he's talking about just checking that `info['lowpass']` and `info['highpass']` are not None for each reader.

**Comment by Eric89GXL** [comment]
(they should default now to 0. and sfreq / 2.)

File: mne/io/tests/test_raw.py [hunk 6]
```diff
@@ -1,14 +1,98 @@
...
+        Raw(out_fname)
+        obj.save(out_fname, tmax=obj.times[-1], overwrite=True)
+        raw3 = Raw(out_fname)
+        assert_equal(sorted(raw.info.keys()), sorted(raw3.info.keys()))
+        assert_array_almost_equal(raw3.load_data()._data[0:20],
+                                  full_data[0:20])
+        assert_array_almost_equal(raw.times, raw3.times)
+
+        assert_true(not math.isnan(raw3.info['highpass']))
+        assert_true(not math.isnan(raw3.info['lowpass']))
```
Comment (jaeilepp) [inline mne/io/tests/test_raw.py]: You mean these?

Comment (jaeilepp) [inline mne/io/tests/test_raw.py]: This needs a fix though...

**Comment by agramfort** [comment]
>
> do you test that highpass and lowpass and consistent between readers?
>
> Didn't we agree to test filtering only for fiff?
>

but it's related to default values. In EGI the bug was due to the reader
using None if not available while the fif reader sets it to 0 and sfreq / 2
if not available.

File: mne/io/tests/test_raw.py (hunk 6 above)
Comment (agramfort) [inline mne/io/tests/test_raw.py]: I added this test as before. EGI was setting highpass and lowpass to None. Then it was written to disk as float hence the Nan when reading back.

File: mne/tests/test_epochs.py (hunk 4 above)
Comment (jaeilepp) [inline mne/tests/test_epochs.py]: Actually, I just removed a redundant creation of epochs object. The decimation here is done to the same object.

File: mne/io/tests/test_raw.py (hunk 6 above)
Comment (jaeilepp) [inline mne/io/tests/test_raw.py]: Okay, seems that some of the readers don't implement this. I'll fix it...

**Comment by jaeilepp** [comment]
Just noticed the CTF class. Maybe that should be refactored as well...

**Comment by Eric89GXL** [comment]
I think all readers under `io/` should be refactored if possible

**Comment by jaeilepp** [comment]
It seems that bti reader allows ``sfreq = None``. Any idea on how to deal with that?

**Comment by Eric89GXL** [comment]
You can end up with a `RawBTi` instance with `raw.info['sfreq'] is None`?

**Comment by jaeilepp** [comment]
It seems that way. Look at bti.py at line 1144.

**Comment by Eric89GXL** [comment]
@dengemann can you comment on this? Looks like you added the `if pdf_fname is None -> raw.info['sfreq'] = None` functionality in one of your BTi edit rounds.

**Comment by jaeilepp** [comment]
I'm changing this back to WIP.

**Comment by Eric89GXL** [comment]
@jaeilepp actually I think @dengemann (correct me if I'm wrong) uses `_get_bti_info` himself with `pdf_fname=None` sometimes, but we probably want `RawBTi` to have a check `if pdf_fname is None: raise ValueError('pdf_fname must be a path, not None')`. Then we're guaranteed that the `RawBTi` instance will have a proper `info['sfreq']` as we require. Make sense?

**Comment by jaeilepp** [comment]
Added ``sfreq`` param to ``empty_info``. EDF class required quite a bit of refactoring to do this.

File: mne/io/bti/bti.py [hunk 7]
```diff
@@ -893,7 +893,8 @@ def _read_bti_header_pdf(pdf_fname):
 def _read_bti_header(pdf_fname, config_fname, sort_by_ch_name=True):
     """ Read bti PDF header
     """
-    info = _read_bti_header_pdf(pdf_fname) if pdf_fname else dict()
+    info = _read_bti_header_pdf(pdf_fname) if pdf_fname else \
+        dict({'sample_period': 0.001})
```
Comment (jaeilepp) [inline mne/io/bti/bti.py]: Had to resort to ugly hack to make the tests pass. Any change we could remove the possibility to have ``sfreq = None``? Ping @dengemann

File: mne/io/brainvision/brainvision.py [hunk 8]
```diff
@@ -358,10 +348,10 @@ def _get_vhdr_info(vhdr_fname, eog, misc, response_trig_shift, scale):
             highpass.append(line[5])
             lowpass.append(line[6])
         if len(highpass) == 0:
-            info['highpass'] = None
+            pass
         elif all(highpass):
             if highpass[0] == 'NaN':
-                info['highpass'] = None
+                pass
```
Comment (jaeilepp) [inline mne/io/brainvision/brainvision.py]: I left these for readability and placeholders if we decide to handle them different in the future.

**Comment by jaeilepp** [comment]
I'm not sure the CTF tests are refactorable without increasing the test times considerably, since the tested data is a bit different. 

@Eric89GXL check out lines 89-90 in ``test_ctf``. I think there is a typo of some sort. ``use_fnames`` is reinitialised right after declaration.

**Comment by Eric89GXL** [comment]
Yeah this line can be deleted:
```
use_fnames = [ctf_eeg_fname, ctf_no_hc_fname]
```

**Comment by Eric89GXL** [comment]
How do you mean the tested data is a bit different? You should at least be able to leave the existing test as is (since it should be fairly complete), and add the `_test_raw` (or whatever it's called) on one of the test files. That might be good enough.

**Comment by jaeilepp** [comment]
You have a lot of fif files you are testing the data against. I can add a test for the reader, but that's just adding a test on top of the rest, not much refactoring.

**Comment by Eric89GXL** [comment]
Yeah that's fine. Just adding is okay for that one.

**Comment by jaeilepp** [comment]
I added test for bti, since it was still missing. Also renamed the test function to something more appropriate.

File: mne/io/bti/bti.py (hunk 7 above)
Comment (jaeilepp) [inline mne/io/bti/bti.py]: Had a short chat with @dengemann . The reason for having None as sfreq is because sometimes it is useful to only read the header info and for bti files you have to read the data to find out the sfreq.

Would it be okay to change it so that in ``_empty_info`` would also accept None as sfreq? Could ``lowpass`` also be None in that case?

**Comment by jaeilepp** [comment]
``_empty_info`` now accepts also None for reading only header info with bti files.

**Comment by Eric89GXL** [comment]
Can we make it so that can only happen when using private functions? In other words, make `read_raw_bti` force `pdf_fname is not None`? It is not good if users end up with an `info['sfreq'] = None`

**Comment by jaeilepp** [comment]
Ready for review.
The test time seems to be 2 seconds longer compared to master for all io tests. CTF and KIT tests in particular contribute to the longer test time.

**Comment by agramfort** [comment]
ok LGTM

just confirm that you can still read the EDF files I shared with you.

great to have so much red lines :)

**Comment by agramfort** [comment]
+1 for merge

@Eric89GXL merge if you're happy.

File: mne/io/meas_info.py [hunk 9]
```diff
@@ -1408,8 +1409,11 @@ def _empty_info():
     for k in _list_keys:
         info[k] = list()
     info['custom_ref_applied'] = False
-    info['nchan'] = info['sfreq'] = 0
+    info['nchan'] = 0
     info['dev_head_t'] = Transform('meg', 'head', np.eye(4))
+    info['highpass'] = 0.
+    info['sfreq'] = sfreq
```
Comment (Eric89GXL) [inline mne/io/meas_info.py]: float(sfreq)

File: mne/io/brainvision/brainvision.py (hunk 8 above)
Comment (Eric89GXL) [inline mne/io/brainvision/brainvision.py]: Thanks for the comment, I was just about to complain :) It would be good to add a comment though because I could imagine in a few months going through here and errantly removing this clause.

Comment (Eric89GXL) [inline mne/io/brainvision/brainvision.py]: Add a comment to the code, I mean

File: mne/io/bti/bti.py [hunk 10]
```diff
@@ -1134,11 +1136,14 @@ def _get_bti_info(pdf_fname, config_fname, head_shape_fname, rotation_x,
 
     use_hpi = False  # hard coded, but marked as later option.
     logger.info('Creating Neuromag info structure ...')
-    info = _empty_info()
+    if 'sample_period' in bti_info.keys():
+        sfreq = 1e3 / bti_info['sample_period'] * 1e-3
```
Comment (Eric89GXL) [inline mne/io/bti/bti.py]: Just noticed this statement does a bit of silly calculation. Should juest be `sfreq = 1. / bti_info['sample_period']`

File: mne/io/edf/edf.py [hunk 11]
```diff
@@ -134,36 +133,33 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
         read_size = len(r_lims) * buf_len
         with open(self._filenames[fi], 'rb', buffering=0) as fid:
             # extract data
-            fid.seek(data_offset +
-                     block_start_idx * buf_len * n_chan * data_size)
+            start_offset = (data_offset +
+                            block_start_idx * buf_len * n_chan * data_size)
+            ch_offsets = np.cumsum(np.concatenate([[0], n_samps])) * data_size
```
Comment (Eric89GXL) [inline mne/io/edf/edf.py]: I know this was my line, but to save one multiply the `* data_size` can go onto `n_samps * data_size` instead of outside the concat and cumsum

File: mne/io/edf/edf.py [hunk 12]
```diff
@@ -449,20 +415,54 @@ def _get_edf_info(fname, stim_channel, annot, annotmap, eog, misc, preload):
...
+    sfreq = n_samps.max() / record_length
+    info = _empty_info(sfreq)
+    info['filename'] = fname
+    info['meas_date'] = calendar.timegm(date.utctimetuple())
+    info['nchan'] = nchan
+    info['chs'] = chs
+    info['ch_names'] = ch_names
+
+    if highpass.size == 0:
+        pass
```
Comment (Eric89GXL) [inline mne/io/edf/edf.py]: same here about leaving a comment in the code

File: mne/io/egi/egi.py [hunk 13]
```diff
@@ -261,10 +261,9 @@ def __init__(self, input_fname, montage=None, eog=None, misc=None,
             # No events
             self.event_id = None
             new_trigger = None
-        info = _empty_info()
+        info = _empty_info(float(egi_info['samp_rate']))
```
Comment (Eric89GXL) [inline mne/io/egi/egi.py]: once the `float` is in `empty_info` you won't need it here

File: mne/io/nicolet/nicolet.py [hunk 14]
```diff
@@ -102,10 +101,11 @@ def _get_nicolet_info(fname, eog, ecg, emg, misc):
     sec, msec = time[2].split('.')
     date = datetime.datetime(int(date[0]), int(date[1]), int(date[2]),
                              int(time[0]), int(time[1]), int(sec), int(msec))
+    info = _empty_info(float(header_info['sample_freq']))
```
Comment (Eric89GXL) [inline mne/io/nicolet/nicolet.py]: nor here

File: mne/io/tests/test_raw.py [hunk 15]
```diff
@@ -1,14 +1,103 @@
...
+            for raw in raws[1:]:
+                data1, times1 = raws[0][picks, sl_time]
+                data2, times2 = raw[picks, sl_time]
+                assert_allclose(data1, data2)
+                assert_allclose(times1, times2)
+
+    raw = raws[-1]  # use preloaded raw
+    full_data = raw._data
+
+    print(raw)  # to test repr
```
Comment (Eric89GXL) [inline mne/io/tests/test_raw.py]: We could do something a bit more clever like `assert_true(raw.__class__.__name__ in repr(raw))` to make sure we're constructing it properly. Same with `raw.info` repr check.

File: mne/io/tests/test_raw.py (hunk 6 above)
Comment (Eric89GXL) [inline mne/io/tests/test_raw.py]: should we also check that they are not None?

File: mne/tests/test_epochs.py [hunk 16]
```diff
@@ -414,16 +414,13 @@ def test_read_write_epochs():
         # decim with lowpass
         warnings.simplefilter('always')
         epochs_dec = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
-                            baseline=(None, 0), decim=4)
+                            baseline=(None, 0), decim=2)
```
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: why the change?

**Comment by Eric89GXL** [comment]
Other than my minor gripes, LGTM! This is long overdue, thanks for tackling it @jaeilepp.

Do you know offhand which readers don't support `preload=False`? We should open an issue about making them all support it. If you don't know offhand I'll look and make the issue.

**Comment by jaeilepp** [comment]
I think bti doesn't. Others i'm not sure.

File: mne/tests/test_epochs.py (hunk 16 above)
Comment (jaeilepp) [inline mne/tests/test_epochs.py]: Well, I started by searching stuff to optimise, and I found this. Basically this just removes one construction of epochs object while staying essentially the same.

**Comment by Eric89GXL** [comment]
Assuming you used your new `_test_raw_reading` function in all our reading classes (did you?), looks like it's just `test_egi` and `test_bti` that say `test_preloading=False`.

File: mne/tests/test_epochs.py (hunk 16 above)
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: Ahh okay cool

**Comment by Eric89GXL** [comment]
Looks like you don't run your new test function to `mne/io/fiff` or `mne/io/array`. Can you add it to those readers for completeness?

**Comment by jaeilepp** [comment]
I thought it would be a bit redundant for fiff.Raw class, since it's basically testing the data against itself, but for RawArray it makes sense.

**Comment by Eric89GXL** [comment]
But your `_test_raw_reader` function doesn't *only* test against the saved/loaded FIFF, it also does other things, like tests that boundaries work properly. That's why I think it should be implemented for FIFF, too.

**Comment by Eric89GXL** [comment]
It's cleaner in my mind to have *all* raw classes be subject to the same set of tests as much as possible, rather than leave one out, even if it means a little bit of redundancy.

**Comment by jaeilepp** [comment]
Okay, I surrender.

**Comment by agramfort** [comment]
>
> Okay, I surrender.
>
you will have your revenge :)

**Comment by Eric89GXL** [comment]
> you will have your revenge

Am I going to be struck by a stale baguette next time I'm over there?

**Comment by agramfort** [comment]
>
> Am I going to be struck by a stale baguette next time I'm over there?
>

something like this :) I guess you prefer baguette than bottle of wine...

**Comment by jaeilepp** [comment]
Just noticed that ``RawArray`` constructor doesn't allow not preloading either.

**Comment by jaeilepp** [comment]
Comments addressed.

**Comment by agramfort** [comment]
I'll let @Eric89GXL merge if he is happy.

nice job @jaeilepp !

**Comment by Eric89GXL** [comment]
Thanks @jaeilepp, more red than green is awesome
